### PR TITLE
WMS-163 | Implement SelectList component

### PR DESCRIPTION
**Pull request by PedroHLC**
Objective:
* Follow [Sufiyan’s design pattern](https://www.sketch.com/s/a252928f-582c-4f38-a5ee-14d60a2a3067/a/zaaeQy) guidelines
* Follow Elm’s with-pattern API guidelines
* Develop one component capable to:
  * [X] Offer a selectable list
  * [x] Action bar
  * [x] Search field

Things that got in together:
* Badge component :name_badge: 
* NavigationContainer's Dialog :smile: 

Things that tried to get-in:
* Text's ellipsis (Not implementable with elm-ui's Text or Paragraph, see: https://github.com/mdgriffith/elm-ui/issues/112)

### CODE REVIEW

File: src/UI/Badge.elm [hunk 1]
```diff
@@ -1,55 +1,58 @@
-module UI.Badge exposing (dark, error, light, primary, success, warning)
+module UI.Badge exposing (Badge, badge, toEl)
```
Comment (jouderianjr) [inline src/UI/Badge.elm:1]: Why did you remove the badge variations? they are there in the design system, check it [here](https://www.sketch.com/s/a252928f-582c-4f38-a5ee-14d60a2a3067/a/G33Qz8)

File: src/UI/Badge.elm [hunk 2]
```diff
@@ -1,55 +1,58 @@
...
 
+toEl : RenderConfig -> Badge -> Element msg
+toEl cfg (Badge { content } { tone }) =
+    Text.overline content
+        |> Text.toEl cfg
+        |> Element.el
+            [ Element.width shrink
+            , Font.center
+            , Background.color Palette.gray.lighter
+            , Element.paddingEach { top = 4, bottom = 4, left = 5, right = 3 }
```
Comment (jouderianjr) [inline src/UI/Badge.elm:46]: left and right padding should normally have the same number, I checked the sketch inspector and it had 5, so I think you could have `paddingXY 5 4`

File: src/UI/SelectList.elm [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+import UI.Icon as Icon exposing (Icon)
+import UI.Internal.Palette as Palette
+import UI.RenderConfig exposing (RenderConfig)
+import UI.Text as Text
+import UI.TextField as TextField
+import UI.Utils.ARIA as ARIA
+import UI.Utils.Element as Element
+
+
+type alias SearchOptions object msg =
```
Comment (jouderianjr) [inline src/UI/SelectList.elm:19]: We currently don't know how this use case will be, so I would remove it until we understand better what we need to do here.

File: src/UI/Button.elm [hunk 4]
```diff
@@ -245,7 +245,7 @@ buttonPadding ((Button { body } _) as btn) =
             Element.paddingXY 31 15
 
         BodyIcon _ ->
-            Element.paddingXY 9 15
+            Element.paddingXY 9 13
```
Comment (jouderianjr) [inline src/UI/Button.elm:248]: Why this was changed?

File: src/UI/Badge.elm (hunk 1 above)
Comment (PedroHLC) [reply src/UI/Badge.elm:1]: WIP

File: src/UI/Badge.elm (hunk 2 above)
Comment (PedroHLC) [reply src/UI/Badge.elm:46]: Letter-spacing is moving text 1px to the left on both Chrome and Firefox for me!

File: src/UI/SelectList.elm (hunk 3 above)
Comment (PedroHLC) [reply src/UI/SelectList.elm:19]: It's already working, on the spaces screen it got something like this, feedbacks are welcomed:
```elm
viewLoadedSpaces : List Space -> Dialog.State Dialog -> Maybe String -> RenderConfig -> Element Msg
viewLoadedSpaces spaces dialog currentFilter cfg =
    selectList (OpenSpaceDetails >> ForDialogMsg) .id spaceItemView
        |> SelectList.withOptions spaces
        |> SelectList.withSearchField
            "Space Name / Barcode"
            FilterSet
            (Maybe.map (\str -> ( str, spaceHasString )) currentFilter)
        |> SelectList.withActionBar "Create New Space"
            Icon.add
            (UuidGen (ForDialogMsg << OpenSpaceCreation))
        |> SelectList.toEl cfg

-- ...
spaceHasString : String -> Space -> Bool
spaceHasString str { name, barcode } =
    String.contains str name
        || String.contains str barcode
```

File: src/UI/Button.elm (hunk 4 above)
Comment (PedroHLC) [reply src/UI/Button.elm:248]: Look at docs, icon-buttons were rendering taller than text button, while on Sketch they're actually smaller (48px vs 40px - was gonna ask you about this anyway)

**Comment by PedroHLC** [comment]
> There are a lot of things here are not related to the pr, can you remove it? (Mainly deletions)

Ok, reverted the deletions, but Badges and Dialog (beside WIP) are too small and one-file change only, I think moving them would only slow the development.

File: src/UI/Button.elm (hunk 4 above)
Comment (PedroHLC) [reply src/UI/Button.elm:248]: Sufyian said 48px of height! This makes it!

**Comment by PedroHLC** [comment]
@jouderianjr, this is back to WIP as I'm writing some tests!

File: src/UI/NavigationContainer.elm [hunk 5]
```diff
@@ -235,3 +291,40 @@ menu applier { menuExpanded } =
...
+        ]
+        [ Element.column
+            [ Element.width (Element.fill |> Element.maximum 640)
+            , Element.centerX
+            , Element.centerY
+            , Background.color <| Palette.gray.lightest
+            ]
+            [ Element.row [ Element.width Element.fill ]
+                [ Text.heading5 title |> Text.toEl cfg |> Element.el [ Element.width Element.fill ]
+                , Icon.close "Close dialog" |> Icon.toEl cfg |> Element.el [ Events.onClick close ]
```
Comment (jouderianjr) [inline src/UI/NavigationContainer.elm:313]: I think we should add `pointer` here

File: src/UI/NavigationContainer.elm [hunk 6]
```diff
@@ -235,3 +291,40 @@ menu applier { menuExpanded } =
 lazyMap : (a -> b) -> (c -> a) -> (c -> b)
 lazyMap applier original =
     \whatever -> applier (original whatever)
+
+
+dialogView : RenderConfig -> Dialog msg -> Element msg
```
Comment (jouderianjr) [inline src/UI/NavigationContainer.elm:296]: Maybe it would be better to move this to another module? Maybe `Internal.Dialog`?

File: src/UI/RenderConfig.elm [hunk 7]
```diff
@@ -37,6 +37,19 @@ fromWindow window =
         |> RenderConfig
 
 
+mapWindow : { window | height : Int, width : Int } -> RenderConfig -> RenderConfig
```
Comment (jouderianjr) [inline src/UI/RenderConfig.elm:40]: this is not exactly a map, right? maybe we could rename to `updateWindowSize`, `setWindowSize` or `updateWindow`.

File: src/UI/SelectList.elm (hunk 3 above)
Comment (jouderianjr) [reply src/UI/SelectList.elm:18]: I said that because we haven't defined if that would be a remote filter, local filter, and so on. it would probably be a remote filter as we will add pagination here. But I'm ok if you prefer to keep this.

File: src/UI/SelectList.elm [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+    { label : String
+    , searchMsg : String -> msg
+    , currentFilter : Maybe ( String, String -> object -> Bool )
+    }
+
+
+type alias Options object msg =
+    -- FUTURE: sort: Maybe (List object -> List object)
+    { itens : List object
```
Comment (jouderianjr) [inline src/UI/SelectList.elm:27]: maybe it's a typo here, would be `items`?

File: src/UI/SelectList.elm [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+    , searchField : Maybe (SearchOptions object msg)
+    , actionBar : Maybe ( String, String -> Icon, msg ) -- TODO
+    , isSelected : Maybe (object -> Bool)
```
Comment (jouderianjr) [inline src/UI/SelectList.elm:30]: I would say that in the middle of the code when I saw `isSelected` I thought it was a `Bool` instead of a function that returns a Bool, so I would suggest calling maybe `isSelectedFn` or something like that because normally we would use `is*` as a Bool, but it's up to you.

File: src/UI/SelectList.elm [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+itemView : RenderConfig -> Properties object msg -> Bool -> object -> Element msg
+itemView cfg { select, renderItem } isSelected obj =
+    let
+        selectedBg : List (Attribute msg) -> List (Attribute msg)
+        selectedBg =
+            Palette.primary.middle
+                |> Background.color
+                |> (::)
+
+        bgColor =
```
Comment (jouderianjr) [inline src/UI/SelectList.elm:192]: So, I'm not a huge fan of point-free style, I think can cause a lot of confusion. mainly because you needed to add a type signature on top to explain what's going on, and we almost never use this inside of `let...in` block.

I think the name of the function is really misleading. a function called, when you think about a function called `bgColor` what it supposes to return?  I would think maybe an `Element.Color` or an `Element.Attribute Msg`, but will never think about returning a list of attributes that are affecting sizing, spacing and etc. and also, it is not select, this function is not even adding a background color.

my suggestion would be

```elm
        selectedBg =
            Background.color Palette.primary.middle
```

and then

```elm
Element.el
    [...attributes
    ]
    |> prependIf isSelected selectedBg

prependIf : Bool -> a -> List a -> List a
prependIf shouldAdd item items =
    if shouldAdd then
         item :: items
    else
         items
```

Ps. this `prependIf` could go in the helper module we discussed before

File: src/UI/Badge.elm [hunk 11]
```diff
@@ -1,55 +1,110 @@
...
+            , Element.height (px 20)
+            , Primitives.roundedBorders
+            ]
+
+
+
+-- Internal
+
+
+toneInside : BadgeTone -> ( Element.Color, TextColor )
```
[the remark this answers is not in the document]
Comment (PedroHLC) [reply src/UI/Badge.elm:91]: Just a helper function for getting color scheme, I'll rename it `toneToColors`, cause that's what it does...

File: src/UI/NavigationContainer.elm (hunk 6 above)
Comment (PedroHLC) [reply src/UI/NavigationContainer.elm:296]: Yep, agreed!

File: src/UI/NavigationContainer.elm (hunk 5 above)
Comment (PedroHLC) [reply src/UI/NavigationContainer.elm:313]: Indeed

File: src/UI/SelectList.elm (hunk 8 above)
Comment (PedroHLC) [reply src/UI/SelectList.elm:27]: It is :portugal: fault!

File: src/UI/SelectList.elm (hunk 9 above)
Comment (PedroHLC) [reply src/UI/SelectList.elm:30]: I prefer `selected` as a boolean, `isSelected` to `sth -> Bool`. We should start writing contribution guidelines...

File: src/UI/SelectList.elm (hunk 10 above)
Comment (PedroHLC) [reply src/UI/SelectList.elm:192]: I liked this `prependIf` a lot, doing it!
